src/index/txindex.cpp

```diff
@@ -57,7 +57,10 @@ bool TxIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
     // Exclude genesis block transaction because outputs are not spendable.
     if (pindex->nHeight == 0) return true;
 
-    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    CDiskTxPos pos{
+        WITH_LOCK(::cs_main, return pindex->GetBlockPos()),
+        GetSizeOfCompactSize(block.vtx.size())
+    };
```

Comment:
0bfa32b131ed2da8799e1d1239bf390dd612b7bb
style nit: [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) insists on:
```suggestion
        GetSizeOfCompactSize(block.vtx.size())};
```

Reply:
> The `WriteUndoDataForBlock` contains disk operations. It seems suboptimal to hold `::cs_main` for the whole function. 

The lock is already held by the function's caller (and is also needed further down in the function for `pindex->nStatus |= BLOCK_HAVE_UNDO;`), so it seemed best to verify the lock is already held rather than taking it twice unnecessarily. WDYT? I updated the commit description to make this more slightly more clear.